examples/hacspec-chacha20/src/chacha20.rs

```diff
@@ -18,7 +18,7 @@ pub fn state_to_bytes(x: State) -> StateBytes {
     r
 }
 
-fn line(a: u32, b: u32, d: u32, s: usize, m: State) -> State {
+fn line(a: StateIdx, b: StateIdx, d: StateIdx, s: usize, m: State) -> State {
```

Comment:
So `StateIdx` is `usize` here? That's fine with me. I just wonder why we can't use `usize` 🙂 .

Reply:
As said above, we need it to prove that the array indexing is correct in F* :)

Reply:
Ideally the Idx type will have dynamic checks in hacspec as well, ensuring that it is within bounds. We don't do it here since Rust does that anyway, I guess?

Reply:
We could wrap the Idx type in a wrapper and implement Numeric for it but it's heavy and the array bounds are already checked by Vec.

Reply:
I just don't like having another argument here for `array` etc. that no one will understand.
And this looks like something we should be able to generate instead of having people write it down explicitly.
If we can't, I guess we have to live with it. But it's not nice 😬

Reply:
We can't generate type identifiers in Rust, it's a fundamental limit of the macron system. However we can change the macro call to something more intelligible, like: 

```
array! (name: State, cell_type: U8, index_type: StateIdx)
```

Reply:
You can actually define your own syntax for whathever is inside the parenthesis.

Reply:
Yeah. I'm afraid you're right. Let's just try to make it as understandable as possible to use 🙂

Reply:
Ok this is the syntax I've implemented : 
```
array!(State, 16, U32, type_for_indexes: StateIdx);
```

Reply:
👍  let's go with this

Reply:
I'll do the same for `nat_mod` and `public_nat_mod`